cannot generate token to model serving

## Summary  

[Provide a brief description of the bug.]  

- cannot generate model serving token

## Steps to Reproduce  

[Outline the steps to consistently reproduce the issue.]  

1. create new serving
1. try to generate a new token
1. you can see the error in manager logs

## Expected Behavior  

[Describe the expected result if the bug were resolved.]  

- successfully generate token

## Actual Behavior  

[Describe what actually happens when the bug occurs.]  

- failed to generate token

## Logs/Errors  

[Attach relevant logs, stack traces, or screenshots to help diagnose the issue.]  

```
lugin.error_monitor [46646] collected an error log [ERROR] "ai.backend.manager.errors.service.EndpointNotFound: No such endpoint. (Failed to generate token: 500 Internal Server Error {'type': 'https://api.backend.ai/probs/internal-server-error', 'title': 'Internal server error.', 'msg': 'Traceback (most recent call last):\n  File "/Users/seungwonlee/Documents/backend.ai/src/ai/backend/appproxy/coordinator/server.py", line 178, in exception_middleware\n    resp = await handler(request)\n           ^^^^^^^^^^^^^^^^^^^^^^\n  File "/Users/seungwonlee/Documents/backend.ai/src/ai/backend/appproxy/coordinator/server.py", line 168, in api_middleware\n    resp = await _handler(request)\n           ^^^^^^^^^^^^^^^^^^^^^^^\n  File "/Users/seungwonlee/Documents/backend.ai/src/ai/backend/common/metrics/http.py", line 39, in metric_middleware\n    resp = await handler(request)\n           ^^^^^^^^^^^^^^^^^^^^^^\n  File "/Users/seungwonlee/Documents/backend.ai/dist/export/python/virtualenvs/python-default/3.13.3/lib/python3.13/site-packages/aiohttp/web_middlewares.py", line 117, in impl\n    return await handler(request)\n           ^^^^^^^^^^^^^^^^^^^^^^\n  File "/Users/seungwonlee/Documents/backend.ai/src/ai/backend/appproxy/coordinator/api/utils.py", line 43, in wrapped\n    return await handler(request, *args, **kwargs)\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File "/Users/seungwonlee/Documents/backend.ai/src/ai/backend/appproxy/common/utils.py", line 240, in wrapped\n    result = await handler(request, checked_params, *args, **kwargs)\n             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File "/Users/seungwonlee/Documents/backend.ai/src/ai/backend/appproxy/coordinator/api/endpoint.py", line 283, in generate_endpoint_api_token\n    payload["app_url"] = str(await circuit.get_endpoint_url())\n                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File "/Users/seungwonlee/Documents/backend.ai/src/ai/backend/appproxy/coordinator/models/circuit.py", line 239, in get_endpoint_url\n    worker: Worker = self.worker_row\n                     ^^^^^^^^^^^^^^^\n  File "/Users/seungwonlee/Documents/backend.ai/dist/export/python/virtualenvs/python-default/3.13.3/lib/python3.13/site-packages/sqlalchemy/orm/attributes.py", line 487, in __get__\n    return self.impl.get(state, dict_)\n           ~~~~~~~~~~~~~^^^^^^^^^^^^^^\n  File "/Users/seungwonlee/Documents/backend.ai/dist/export/python/virtualenvs/python-default/3.13.3/lib/python3.13/site-packages/sqlalchemy/orm/attributes.py", line 959, in get\n    value = self._fire_loader_callables(state, key, passive)\n  File "/Users/seungwonlee/Documents/backend.ai/dist/export/python/virtualenvs/python-default/3.13.3/lib/python3.13/site-packages/sqlalchemy/orm/attributes.py", line 995, in _fire_loader_callables\n    return self.callable_(state, passive)\n           ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^\n  File "/Users/seungwonlee/Documents/backend.ai/dist/export/python/virtualenvs/python-default/3.13.3/lib/python3.13/site-packages/sqlalchemy/orm/strategies.py", line 912, in _load_for_state\n    return self._emit_lazyload(\n           ~~~~~~~~~~~~~~~~~~~^\n        session,\n        ^^^^^^^^\n    ...<4 lines>...\n        extra_criteria,\n        ^^^^^^^^^^^^^^^\n    )\n    ^\n  File "/Users/seungwonlee/Documents/backend.ai/dist/export/python/virtualenvs/python-default/3.13.3/lib/python3.13/site-packages/sqlalchemy/orm/strategies.py", line 996, in _emit_lazyload\n    return loading.load_on_pk_identity(\n           ~~~~~~~~~~~~~~~~~~~~~~~~~~~^\n        session, stmt, primary_key_identity, load_options=load_options\n        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n    )\n    ^\n  File "/Users/seungwonlee/Documents/backend.ai/dist/export/python/virtualenvs/python-default/3.13.3/lib/python3.13/site-packages/sqlalchemy/orm/loading.py", line 530, in load_on_pk_identity\n    session.execute(\n    ~~~~~~~~~~~~~~~^\n        q,\n        ^^\n    ...<2 lines>...\n        bind_arguments=bind_arguments,\n        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n    )\n    ^\n  File "/Users/seungwonlee/Documents/backend.ai/dist/export/python/virtualenvs/python-default/3.13.3/lib/python3.13/site-packages/sqlalchemy/orm/session.py", line 1717, in execute\n    result = conn._execute_20(statement, params or {}, execution_options)\n  File "/Users/seungwonlee/Documents/backend.ai/dist/export/python/virtualenvs/python-default/3.13.3/lib/python3.13/site-packages/sqlalchemy/engine/base.py", line 1710, in _execute_20\n    return meth(self, args_10style, kwargs_10style, execution_options)\n  File "/Users/seungwonlee/Documents/backend.ai/dist/export/python/virtualenvs/python-default/3.13.3/lib/python3.13/site-packages/sqlalchemy/sql/elements.py", line 334, in _execute_on_connection\n    return connection._execute_clauseelement(\n           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^\n        self, multiparams, params, execution_options\n        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n    )\n    ^\n  File "/Users/seungwonlee/Documents/backend.ai/dist/export/python/virtualenvs/python-default/3.13.3/lib/python3.13/site-packages/sqlalchemy/engine/base.py", line 1577, in _execute_clauseelement\n    ret = self._execute_context(\n        dialect,\n    ...<8 lines>...\n        cache_hit=cache_hit,\n    )\n  File "/Users/seungwonlee/Documents/backend.ai/dist/export/python/virtualenvs/python-default/3.13.3/lib/python3.13/site-packages/sqlalchemy/engine/base.py", line 1953, in _execute_context\n    self._handle_dbapi_exception(\n    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^\n        e, statement, parameters, cursor, context\n        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n    )\n    ^\n  File "/Users/seungwonlee/Documents/backend.ai/dist/export/python/virtualenvs/python-default/3.13.3/lib/python3.13/site-packages/sqlalchemy/engine/base.py", line 2138, in _handle_dbapi_exception\n    util.raise_(exc_info[1], with_traceback=exc_info[2])\n    ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File "/Users/seungwonlee/Documents/backend.ai/dist/export/python/virtualenvs/python-default/3.13.3/lib/python3.13/site-packages/sqlalchemy/util/compat.py", line 211, in raise_\n    raise exception\n  File "/Users/seungwonlee/Documents/backend.ai/dist/export/python/virtualenvs/python-default/3.13.3/lib/python3.13/site-packages/sqlalchemy/engine/base.py", line 1910, in _execute_context\n    self.dialect.do_execute(\n    ~~~~~~~~~~~~~~~~~~~~~~~^\n        cursor, statement, parameters, context\n        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n    )\n    ^\n  File "/Users/seungwonlee/Documents/backend.ai/dist/export/python/virtualenvs/python-default/3.13.3/lib/python3.13/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute\n    cursor.execute(statement, parameters)\n    ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^\n  File "/Users/seungwonlee/Documents/backend.ai/dist/export/python/virtualenvs/python-default/3.13.3/lib/python3.13/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 479, in execute\n    self._adapt_connection.await_(\n    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^\n        self._prepare_and_execute(operation, parameters)\n        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n    )\n    ^\n  File "/Users/seungwonlee/Documents/backend.ai/dist/export/python/virtualenvs/python-default/3.13.3/lib/python3.13/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 61, in await_only\n    raise exc.MissingGreenlet(\n    ...<2 lines>...\n    )\nsqlalchemy.exc.MissingGreenlet: greenlet_spawn has not been called; can\'t call await_only() here. Was IO attempted in an unexpected place? (Background on this error at: https://sqlalche.me/e/14/xd2s)\n'})" from manager
```


JIRA Issue: BA-2083

--- a/changes/5448.fix.md
+++ b/changes/5448.fix.md
@@ -1,0 +1,1 @@
+Change `Circuit` query logic in AppProxy for `Circuit` object's serializability

--- a/src/ai/backend/appproxy/coordinator/api/endpoint.py
+++ b/src/ai/backend/appproxy/coordinator/api/endpoint.py
@@ -276,11 +276,11 @@ async def generate_endpoint_api_token(
 
     async with root_ctx.db.begin_readonly_session() as sess:
         circuit: Circuit = await Circuit.find_by_endpoint(
-            sess, UUID(request.match_info["endpoint_id"]), load_worker=False
+            sess, UUID(request.match_info["endpoint_id"]), load_worker=False, load_endpoint=False
         )
         payload = dict(circuit.dump_model())
         payload["config"] = {}
-        payload["app_url"] = str(await circuit.get_endpoint_url())
+        payload["app_url"] = str(await circuit.get_endpoint_url(session=sess))
 
     payload["user"] = str(params.user_uuid)
     payload["exp"] = params.exp

--- a/src/ai/backend/appproxy/coordinator/models/circuit.py
+++ b/src/ai/backend/appproxy/coordinator/models/circuit.py
@@ -1,5 +1,5 @@
 from datetime import datetime
-from typing import TYPE_CHECKING, Any, Sequence
+from typing import TYPE_CHECKING, Any, Optional, Sequence
 from uuid import UUID
 
 import sqlalchemy as sa
@@ -233,8 +233,11 @@ class Circuit(Base, BaseMixin):
 
         return c
 
-    async def get_endpoint_url(self) -> URL:
+    async def get_endpoint_url(self, session: Optional[AsyncSession] = None) -> URL:
         from .worker import Worker
+
+        if session is not None:
+            self.worker_row = await Worker.get(session, self.worker)
 
         worker: Worker = self.worker_row
         match (worker.use_tls, self.protocol):

--- a/src/ai/backend/appproxy/coordinator/models/circuit.py
+++ b/src/ai/backend/appproxy/coordinator/models/circuit.py
@@ -236,10 +236,10 @@ class Circuit(Base, BaseMixin):
     async def get_endpoint_url(self, session: Optional[AsyncSession] = None) -> URL:
         from .worker import Worker
 
-        if session is not None:
-            self.worker_row = await Worker.get(session, self.worker)
+        worker: Worker = (
+            await Worker.get(session, self.worker) if session is not None else self.worker_row
+        )
 
-        worker: Worker = self.worker_row
         match (worker.use_tls, self.protocol):
             case (True, ProxyProtocol.TCP):
                 scheme = "tls"

